File: incre-tests/synduce/constraints/sortedlist/nestedl_pyramid.f
Inductive List = elt Int | cons {Int, List};
Inductive NList = line List | ncons {List, NList};

max = \a: Int. \b: Int. if < a b then b else a;
lmax = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> max h (f t)
  end
);

min = \a: Int. \b: Int. if < a b then a else b;
lmin = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> min h (f t)
  end
);

is_sorted =
  let aux = fix (
    \f: Int -> NList -> Bool. \pre: Int. \xs: NList.
    match xs with
      line x -> <= pre (lmax x)
    | ncons {h, t} -> and (<= pre (lmax h)) (f (lmax h) t)
    end
  ) in \xs: NList.
  match xs with
    line _ -> true
  | ncons {h, t} -> aux (lmax h) t
  end;

interval = fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    elt x -> {x, x}
  | cons {h, t} ->
    let res = f t in
      {min res.1 h, max res.2 h}
  end
);

spec = \xs: NList. (fix (
  \f: NList -> {Int, Int, Bool}. \xs: NList.
  match xs with
    line x ->
      let res = interval x in
        {res.1, res.2, true}
  | ncons {h, t} ->
      let info = interval h in
        let res = f t in
          {min info.1 res.1, max info.2 res.2,
           and res.3 (and (<= res.1 info.1) (>= res.2 info.2))}
  end
) xs).3;

target = fix (
  \f: NList -> Compress NList. \xs: NList.
  match xs with
    line x ->
      let info = interval x in /*This invocation is provided in Synduce's template*/
        xs
  | ncons {h, t} ->
      let mi = lmin h in /*This invocation is provided in Synduce's template*/
        ncons {h, f t}
   end
);

main = \xs: NList. if is_sorted xs then spec (target xs) else false;

File: ml-benchmark/synduce/constraints/sortedlist/nestedl_pyramid.ml
type list = Elt of int | Cons of int * list
type nlist = Line of list | Ncons of list * nlist

let max a b = if a < b then b else a

let rec lmax xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) -> max h (lmax t)

let min a b = if a < b then a else b

let rec lmin xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) -> min h (lmin t)

let is_sorted xs =
  let rec aux pre xs =
    match xs with
    | Line x -> pre <= lmax x
    | Ncons (h, t) -> (pre <= lmax h) && aux (lmax h) t
  in
  match xs with
  | Line _ -> true
  | Ncons (h, t) -> aux (lmax h) t

let rec interval xs =
  match xs with
  | Elt x -> (x, x)
  | Cons (h, t) ->
      let res = interval t in
      match res with
      | (a, b) -> (min a h, max b h)

let rec spec_helper xs =
  match xs with
  | Line x ->
      let res = interval x in
      match res with
      | (a, b) -> (a, b, true)
  | Ncons (h, t) ->
      let info = interval h in
      let res = spec_helper t in
      match info with
      | (i1, i2) ->
          match res with
          | (r1, r2, r3) ->
              (min i1 r1, max i2 r2, r3 && ((r1 <= i1) && (r2 >= i2)))

let spec xs =
  let res = spec_helper xs in
  match res with
  | (_, _, b) -> b

let rec target xs =
  match xs with
  | Line x ->
      let info = interval x in
      xs
  | Ncons (h, t) ->
      let mi = lmin h in
      Ncons (h, target t)

let program xs =
  if is_sorted xs then spec (target xs) else false